walk/walk.go

```diff
@@ -122,25 +124,26 @@ func Walk(c *config.Config, cexts []config.Configurer, dirs []string, mode Mode,
 		log.Printf("error loading .bazelignore: %v", err)
 	}
 
-	visit(c, cexts, isBazelIgnored, knownDirectives, updateRels, wf, c.RepoRoot, "", false)
-}
-
-func visit(c *config.Config, cexts []config.Configurer, isBazelIgnored isIgnoredFunc, knownDirectives map[string]bool, updateRels *UpdateFilter, wf WalkFunc, dir, rel string, updateParent bool) {
-	if isBazelIgnored(rel) {
-		return
+	trie, err := buildTrie(c, isBazelIgnored)
+	if err != nil {
+		log.Fatalf("error walking the file system: %v\n", err)
 	}
 
+	visit(c, cexts, knownDirectives, updateRels, trie, wf, c.RepoRoot, "", false)
+}
+
+func visit(c *config.Config, cexts []config.Configurer, knownDirectives map[string]bool, updateRels *UpdateFilter, trie *pathTrie, wf WalkFunc, dir, rel string, updateParent bool) {
 	haveError := false
 
-	// TODO: OPT: ReadDir stats all the files, which is slow. We just care about
-	// names and modes, so we should use something like
-	// golang.org/x/tools/internal/fastwalk to speed this up.
-	ents, err := os.ReadDir(dir)
-	if err != nil {
-		log.Print(err)
-		return
+	ents := make([]fs.DirEntry, 0, len(trie.children))
```

Comment:
Nit: I would probably avoid the 0 arg and do 'ents[I] = node.entry' in the loop, it's a bit shorter.

Reply:
That's not so easy because `children` is a map and we're building an array

Reply:
That's also the reason we must sort it here instead of in the parallel fs walk.

Reply:
Oh, right. Sorry